src/java.base/share/classes/sun/nio/ch/ChannelOutputStream.java

```diff
@@ -64,10 +64,16 @@ WritableByteChannel channel() {
      * If the channel is selectable then it must be configured blocking.
      */
     private void writeFully(ByteBuffer bb) throws IOException {
-        while (bb.remaining() > 0) {
+        int pos = bb.position();
+        int rem = bb.limit() - pos;
+        while (rem > 0) {
+            bb.limit(pos + Math.min(Streams.MAX_BUFFER_SIZE, rem));
             int n = ch.write(bb);
             if (n <= 0)
-                throw new RuntimeException("no bytes written");
+                throw new IOException("Write failed");
+            pos += n;
+            rem -= n;
+            bb.position(pos);
```

Comment:
I assume this line is not needed as ch.write(bb) will advance the position by n.